src/content/en/fundamentals/media/manipulating/live-effects.md

```diff
@@ -0,0 +1,376 @@
+
+    // Draw the image onto the top-left corner of the canvas
+    context.drawImage(theOriginalImage, 0, 0);
+
+Then you get an array of pixel values for the whole canvas.
+
+    const imageData = context.getImageData(0, 0, canvas.width, canvas.height);
+    const pixels = imageData.data;
+
+At this point the `pixels` variable is a `Uint8ClampedArray` with a length of `width * height * 4`.
+The data is arranged so that every pixel in the image, starting at the top-left and working left to
+right, top to bottom, has 4 bytes of data. The bytes represent the amounts of red, green and blue as
+well as the alpha, or transparency.
+
+    pixels[0] = red value for pixel 0
+    pixels[1] = green value for pixel 0
+    pixels[2] = blue value for pixel 0
+    pixels[3] = alpha value for pixel 0
+    pixels[4] = red value for pixel 1
+    pixels[5] = green value for pixel 1
+    pixels[6] = blue value for pixel 1
+    pixels[7] = alpha value for pixel 1
+    pixels[8] = red value for pixel 2
+    pixels[9] = green value for pixel 2
+    pixels[10] = blue value for pixel 2
+    pixels[11] = alpha value for pixel 2
+    pixels[12] = red value for pixel 3
+    ...
+
+To find the indexes for any given pixel from its coordinates, there is a simple formula.
+
+    const index = (x + y * imageWidth) * 4;
+    const red = pixels[index];
+    const green = pixels[index + 1];
+    const blue = pixels[index + 2];
+    const alpha = pixels[index + 3];
+
+You can now read and write this data however you want, allowing you to apply any effects you can
+think of. However, this array is a *copy* of the actual pixel data for the canvas. To write the
+edited version back you need to use the `putImageData` method to write this back to the top-left
+corner of the canvas
+
+    context.putImageData(imageData, 0, 0);
+
+### WebGL
+Anyone who has worked with WebGL before will probably realise that I can't go into the full detail
+of how this works in a single article. If you'd like to learn more about WebGL, check out the
+recommended reading at the end of this article.
+
+However, you can take a general overview of what needs to be done in the case of manipulating a
+single image.
+
+One of the most important things to remember about WebGL is that it is **not** a 3D graphics API. In
+fact, WebGL (and OpenGL) is good at exactly one thing - drawing triangles. In your application you
+must work out describe what you *actually* want to draw in terms of triangles. In the case of a 2D
+image, that's very simple, because a rectangle is two similar right-angled triangles, arranged so
+that their hypotenuses are in the same place.
+
+<img src="images/quad-from-triangles.png" class="attempt-right" />
+
+The basic process is:
+- Send data to the GPU that describes the vertices (points) of the triangles
+- Send your source image to the GPU as a texture (image)
+- Create a 'vertex shader'
+- Create a 'fragment shader'
+- Set some shader variables, called 'uniforms'
+- Run the shaders
```

Comment:
Not strictly ordered. The last item must be last, but the first two can happen anywhere before that. Setting uniforms must be after creating the shaders, but the shaders can be done either way around.